lib/app/modules/us_core_r4/clinicalnotes_sequence.rb

```diff
@@ -0,0 +1,228 @@
+
+        assert @must_have_type_code[type_code], "Clinical Notes shall have at least one DocumentReference with type #{type_code}"
+      end
+
+      def test_clinicalnotes_diagnosticreport(category_code)
+        skip 'No Clinical Notes appear to be available for this patient. Please use patients with more information.' unless @clinicalnotes_found
+
+        patient_val = @instance.patient_id
+        search_params = { 'patient': patient_val, 'category': category_code }
+
+        reply = get_resource_by_params(versioned_resource_class('DiagnosticReport'), search_params)
+        assert_response_ok(reply)
+        assert_bundle_response(reply)
+
+        resource_count = reply&.resource&.entry&.length || 0
+        @resources_found = true if resource_count.positive?
+
+        assert @resources_found, "Clinical Notes shall have at least one DiagnosticReport with category #{category_code}"
+
+        diagnosticreport_ary = reply&.resource&.entry&.map { |entry| entry&.resource }
+
+        diagnosticreport_ary&.each do |resource|
+          id = resource&.id
+
+          resource&.presentedForm&.each do |attachment|
+            url_value = attachment&.url
+            next if url_value.nil?
+
+            assert @attachment_url.key?(url_value), "Attachment #{url_value} referenced in DiagnosticReport/#{id} but not in any DocumentReference"
+            @attachment_url[url_value][:flag] = true
+          end
+        end
+      end
+
+      test 'Server returns expected results from DocumentReference search by patient+clinicalnotes' do
+        metadata do
+          id '01'
+          link 'https://build.fhir.org/ig/HL7/US-Core-R4/clinical-notes-guidance.html'
+          desc %(
+          )
+          versions :r4
+        end
+
+        patient_val = @instance.patient_id
+        category_val = 'clinical-note'
+        search_params = { 'patient': patient_val, 'category': category_val }
+        @attachment_url = {}
+
+        reply = get_resource_by_params(versioned_resource_class('DocumentReference'), search_params)
+        assert_response_ok(reply)
+        assert_bundle_response(reply)
+
+        resource_count = reply&.resource&.entry&.length || 0
+        @clinicalnotes_found = true if resource_count.positive?
+
+        skip 'No Clinical Notes appear to be available for this patient. Please use patients with more information.' unless @clinicalnotes_found
+
+        documentreference_ary = reply&.resource&.entry&.map { |entry| entry&.resource }
+
+        @must_have_type_code = {
+          '11488-4' => false, # Consultation Note
+          '18842-5' => false, # Dischard Summary
+          '34117-2' => false, # History and physical note
+          '28570-0' => false, # Procedure note
+          '11506-3' => false # Progress not
+        }
+
+        documentreference_ary&.each do |resource|
+          codings = resource&.type&.coding
+          next if codings.nil?
```

Comment:
When possible, perform some sort of filter operation rather than using `next` inside of `each`. In this case I'd probably do something like:
```
document_references
  .select { |document_reference| document_reference&.resource&.type&.coding&.present? }
  .each do |document_reference|
    ...
```
The nested eaches could do something similar.

Reply:
fixed